Support Digital level with software I2C

ClutchplateDude opened the pull request.

ClutchplateDude committed:
MPU6050 Updates

- Support Software I2C library for MPU6050 chip.

ClutchplateDude committed:
Updates

- Gyro code fixed to read 16-bit ints a byte at a time correctly.

ClutchplateDude committed:
V1.9.15 - Updates

- Add support for software I2C so that Digital Level can be used on MKS board.

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -5,27 +5,35 @@
- * Datasheet: https://invensense.tdk.com/wp-content/uploads/2015/02/MPU-6000-Datasheet1.pdf
- * Register descriptions: https://invensense.tdk.com/wp-content/uploads/2015/02/MPU-6000-Register-Map1.pdf
- * */
+// Tilt, roll, and temperature measurementusing the MPU-6050 MEMS gyro.
```
Try not to change comment styles with no additional changes, it just adds code churn

ClutchplateDude replied:
You know how you say always leave code better than how you find it? enums vs. typedefs, etc. coding style is one of those. :-)

julianneswinoga replied:
Yeah I know, it's a fine line :)
My reasoning:
- There are no functional benefits
- It's subjective (for now, until we have coding guidelines)

ClutchplateDude replied:
The style is pretty consistent (with expections), but nowhere do we use this style of commenting...

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -5,27 +5,35 @@
+// Static var to keep track of whether the MPU6050 was found
+bool Gyro::isPresent = false;
 
+// Starts up the MPU-6050 device.
```
Code churn here too

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -34,67 +42,88 @@ void Gyro::startup()
     Wire.endTransmission(true);
     Wire.requestFrom(MPU6050_I2C_ADDR, 1, 1);
-    byte id = (Wire.read() >> 1) & 0x3F;    
+    byte id = Wire.read();
```
```C++
const byte id = ((Wire.read() >> 1) & 0x3F);
isPresent = (id == 0x34);
```
?
There are wayyyy better ways to do register reading/comparisons but this can still be cleaned up a tad

ClutchplateDude replied:
This used to be on one line, but I split it for debugging purposes.... is there a request here?

julianneswinoga replied:
In my opinion it's a much cleaner implementation, shows the intent a lot better, const-correctness, etc. Up to you in the end though :upside_down_face:

ClutchplateDude replied:
Oh, I didn't realize that was new code you wrote, lol. Like I said while trying to get this running, I wanted to see what actually was read before the bit fiddling.. But now that it works, I'll change it to the more optimized now.

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -34,67 +42,88 @@ void Gyro::startup()
     Wire.endTransmission(true);
 
-    // Execute 1 byte write to MPU6050_REG_PWR_MGMT_1
+    // Execute 1 byte read/write to MPU6050_REG_ACCEL_CONFIG to set 4g sensititvity
```
Spelling `sensitivity`

ClutchplateDude replied:
Fixed

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -34,67 +42,88 @@ void Gyro::startup()
+    Wire.endTransmission();
+    Wire.requestFrom(MPU6050_I2C_ADDR, 1);
+    byte x = Wire.read(); //the value of Register-28 is in x
+    x = (x & 0b11100111);     //clear values of Bit4 and Bit3 fo 2g sensitivity
```
You say 4g sensitivity above, which one is it?

ClutchplateDude replied:
Fixed

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -34,67 +42,88 @@ void Gyro::startup()
+    Wire.write(MPU6050_REG_ACCEL_CONFIG);
+    Wire.endTransmission();
+    Wire.requestFrom(MPU6050_I2C_ADDR, 1);
+    byte x = Wire.read(); //the value of Register-28 is in x
```
```C++
const byte accelCfgReg = (Wire.read() & 0b11100111);
```
?

ClutchplateDude replied:
Yup.

julianneswinoga replied:
I meant the code line, same as the other register refactor above

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -34,67 +42,88 @@ void Gyro::startup()
         result.rollAngle += ((atanf(-1 * AcY / sqrtf(powf(AcX, 2) + powf(AcZ, 2))) * 180.0f / static_cast<float>(PI)) * 2.0f) / 2.0f;
 
-        delay(10);  // Decorrelate measurements
+        delay(5); // Decorrelate measurements
```
Why change the update rate + window size here? In the initialization you set 5Hz bandwidth but here you're checking at 200Hz (1/(5ms) == 200Hz). Am I missing something?

ClutchplateDude replied:
Not sure. Maybe I'm not understanding how this works.

ClutchplateDude replied:
Fixed.

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -34,67 +42,88 @@ void Gyro::startup()
+        Wire.write(MPU6050_REG_ACCEL_XOUT_H);
         Wire.endTransmission(false);
-        Wire.requestFrom(MPU6050_I2C_ADDR, 6, 1);  // Read 6 registers total, each axis value is stored in 2 registers
-        int16_t AcX = Wire.read() << 8 | Wire.read(); // X-axis value
```
Mmm yeah this was sketchy, gj breaking it up :+1:

ClutchplateDude replied:
Yeah, no guarantee in what order the compiler calls read

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -34,67 +42,88 @@ void Gyro::startup()
-        int16_t AcY = Wire.read() << 8 | Wire.read(); // Y-axis value
-        int16_t AcZ = Wire.read() << 8 | Wire.read(); // Z-axis value
+        Wire.requestFrom(MPU6050_I2C_ADDR, 6, 1);     // Read 6 registers total, each axis value is stored in 2 registers
+        uint8_t b0 =  Wire.read();
```
I'd rename to `AxXLow`/`AcXHigh` or something to be more explicit, but up to you. (I would also pepper `const` everywhere but that's just me :grin: )

ClutchplateDude replied:
Fixed

julianneswinoga reviewed (changes requested):
@andre-stefanov @ClutchplateDude @CTetford 
In the future we should look into using git submodules instead of raw source clones... It would guarantee the git hash and be easy to update.

ClutchplateDude committed:
PR Feedback fixes

ClutchplateDude committed:
Merge branch 'develop' into i2c-level

ClutchplateDude committed:
Bumped version

ClutchplateDude committed:
More PR feedback updates

ClutchplateDude committed:
More Updates

- Added Gyro pins or comments to boards
- Improved the display of pitch and roll in the LCD.
- Removed some smoothing from the gyro calls. More smoothing is left to the caller.
- Freed up 600 bytes of RAM by moving some strings to Flash.

ClutchplateDude committed:
Gyro Updates

Commented out the timeouts. Seems we need another library for that.

julianneswinoga commented on boards/AVR_MKS_GEN_L_V1/pins_MKS_GEN_L_V1.h:
```diff
@@ -150,6 +150,8 @@
   #define ALT_DRIVER_ADDRESS 0b00
 #endif
 
+// Digital Level addon - not officially supported on ESP32. If you find two free pins, add GYRO_SOFTWARE_SCL_PIN  and GYRO_SOFTWARE_SDA_PIN to your local config
```
Shouldn't this be MKS GenL V1 instead of ESP32?

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -5,100 +5,133 @@
+    Wire.write(MPU6050_REG_ACCEL_CONFIG);
+    Wire.endTransmission();
+    Wire.requestFrom(MPU6050_I2C_ADDR, 1);
+    byte accelConfigReg = Wire.read();              //the value of Register-28 is in x
```
Could you do the same thing here as I requested above?
```C++
const byte accelConfigReg = (Wire.read() & 0b11100111);
```

ClutchplateDude replied:
Done

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -5,100 +5,133 @@
+    result.rollAngle += ((atanf(-1 * accelInY / sqrtf(powf(accelInX, 2) + powf(accelInZ, 2))) * 180.0f / static_cast<float>(PI)) * 2.0f) / 2.0f;
+
+#if USE_GYRO_WITH_SOFTWAREI2C == 0
+    // if (Wire.getWireTimeoutFlag())
```
Delete unused debug code

ClutchplateDude replied:
This is not debug code. There is a library that supports timeouts

julianneswinoga replied:
Why is it commented out then? As far as I'm concerned, if code is commented out it's debug.

ClutchplateDude replied:
There is a library upgrade coming where this will be used. I don't want to have to think about getting this code back once it's included.

julianneswinoga replied:
Ideally we would implement that code when the functionality is needed, but I can understand why you want to leave it now. Can you at least add a `// TODO` comment explaining why the code is unused?

julianneswinoga commented on src/Gyro.cpp:
```diff
@@ -5,100 +5,133 @@
+    Wire.setClock(100000); // Set lowest clock speed (100kHz) to make the communication more reliable
 
+#if USE_GYRO_WITH_SOFTWAREI2C == 0
+    // Wire.setWireTimeout(3000, true); // Set I2C timeout if the Wire.h calls are hanging due to a HW issue
```
Delete unused debug code

ClutchplateDude replied:
Ditto

julianneswinoga commented on src/c76_menuCAL.hpp:
```diff
@@ -1231,28 +1248,46 @@ void printCalibrationSubmenu()
+      accumulator.rollAngle /= (float)ACCUMULATOR_COUNT;
+      sprintf(scratchBuffer, "R: ");
+      makeIndicator(scratchBuffer, accumulator.rollAngle - mount.getRollCalibrationAngle());
+      scratchBuffer[16] = 0;
```
Use `'\0'` instead of `0` for null byte in string

ClutchplateDude replied:
Fixed.

julianneswinoga commented on src/c76_menuCAL.hpp:
```diff
@@ -1231,28 +1248,46 @@ void printCalibrationSubmenu()
+    accumulator.rollAngle += angles.rollAngle;
+    if (calCount >= ACCUMULATOR_COUNT)
+    {
+      accumulator.rollAngle /= (float)ACCUMULATOR_COUNT;
```
Could you instead implement the details in the gyro module instead of the menu file? It doesn't make sense to do this here.

Also while we're changing this, doing this dance with a calibration counter, accumulator, reset, etc seems like a very simplistic way to implement a low pass filter. Something like an exponential low pass filter ([example C implementation](https://kiritchatterjee.wordpress.com/2014/11/10/a-simple-digital-low-pass-filter-in-c/)) is very simple to implement and would work fantastic for our use case. :smile: 
I can definitely help you out here if you need it!

ClutchplateDude replied:
I don't want the Gyro class to do this, it would make it unneccesarily slower. But here the frequent LCD updates make it unreadable. I want it a lot slower, but only for the LCD. So this is the right place. Won't fix.

ClutchplateDude replied:
OK, that low pass filter works really well (in the test app). Trying to see if I can get it to behave in the firmware

andre-stefanov commented on src/Gyro.cpp:
```diff
@@ -5,100 +5,133 @@
+
+    // Execute 1 byte write to MPU6050_REG_CONFIG to set slowest refresh rate
     Wire.beginTransmission(MPU6050_I2C_ADDR);
     Wire.write(MPU6050_REG_CONFIG);
-    Wire.write(6);      // 5Hz bandwidth (lowest) for smoothing
+    Wire.write(6); // 5Hz bandwidth (lowest) for smoothest readings
     Wire.endTransmission(true);
```
you could bundle these calls to a function something like `send(reg, val)`

andre-stefanov replied:
even with default value `null` for val if only register has to be sent

ClutchplateDude committed:
Added low pass filter

julianneswinoga commented on src/c76_menuCAL.hpp:
```diff
@@ -1232,27 +1249,36 @@ void printCalibrationSubmenu()
-    sprintf(scratchBuffer, "R: -------------");
-    makeIndicator(scratchBuffer, angles.rollAngle - mount.getRollCalibrationAngle());
-    lcdMenu.printMenu(scratchBuffer);
+    float delta = accumulator.rollAngle - angles.rollAngle;
```
I still think that this is not the place for the calculations, they should be abstracted in the gyro interface such that the menu code doesn't care about how the calculations work. The menu just wants the angles and the gyro takes care of how that happens, which reduces module coupling (https://en.wikipedia.org/wiki/Coupling_(computer_programming))

ClutchplateDude replied:
I disagree. As a consumer of the class, I might not want it smoothed. I might want a different smoothing factor. I might be calling it irregularly and smoothing would be worse....

andre-stefanov replied:
Then it should be a parameter on gyro api. But here menu is changing actual behaviour while it should only trigger a feature without knowing the actual implementation.

Imagine we want to provide this as meade as well.

julianneswinoga replied:
> As a consumer of the class, I might not want it smoothed. I might want a different smoothing factor.

Simple, two functions then: `Gyro::getCurrentAnglesRaw(void)` and `Gyro::getCurrentAnglesFiltered(float smoothFactor)`

> I might be calling it irregularly and smoothing would be worse

That's up to the consumer of the class to use the interface correctly. You can fix that in the gyro interface with more complexity (threading, etc) but that's not needed for simple usage like this.

julianneswinoga commented on src/c76_menuCAL.hpp:
```diff
@@ -94,6 +94,12 @@ enum
+// Variables to accumulate and smooth the accelerometer readings of the digital level
+#if USE_GYRO_LEVEL == 1
+angle_t accumulator;
+float LPFBeta = 0.6f;
```
You should get in the habit of using `const` wherever possible: https://isocpp.org/wiki/faq/const-correctness

ClutchplateDude replied:
Sure.

ClutchplateDude commented:
Abandoning this PR. Digital Level is very underused and code change is too old.
